private/buf/cmd/buf/command/push/push.go

```diff
@@ -347,6 +386,155 @@ func validateLabelFlagValues(flags *flags) error {
+	if err != nil {
+		// We surface additional information here for the user to ensure they are using this flag
+		// with a git checkout.
+		return nil, fmt.Errorf("unable to check input %q, please ensure this is a Git repository checkout: %w", input, err)
+	}
+	if len(uncommittedFiles) > 0 {
+		return nil, fmt.Errorf("uncommitted changes found in the following files: %v", uncommittedFiles)
+	}
+	originRemote, err := git.GetRemote(ctx, runner, app.EnvironMap(container), input, gitOriginRemote)
+	if err != nil {
+		if errors.Is(err, git.ErrRemoteNotFound) {
+			return nil, appcmd.NewInvalidArgumentErrorf("remote %s must be present on Git checkout: %w", gitOriginRemote, err)
+		}
+		return nil, err
+	}
+	currentGitCommit, err := git.GetCurrentHEADGitCommit(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
+	var gitMetadataUploadOptions []bufmodule.UploadOption
+	gitLabelsUploadOption, err := getGitMetadataLabelsUploadOptions(ctx, runner, app.EnvironMap(container), input, currentGitCommit)
+	if err != nil {
+		return nil, err
+	}
+	if gitLabelsUploadOption != nil {
+		gitMetadataUploadOptions = append(gitMetadataUploadOptions, gitLabelsUploadOption)
+	}
+	gitSourceControlURLUploadOption, err := getGitMetadataSourceControlURLUploadOption(originRemote, currentGitCommit)
+	if err != nil {
+		return nil, err
+	}
+	if gitSourceControlURLUploadOption != nil {
+		gitMetadataUploadOptions = append(gitMetadataUploadOptions, gitSourceControlURLUploadOption)
+	}
+	if flags.Create {
+		createModuleVisibility, err := bufmodule.ParseModuleVisibility(flags.CreateVisibility)
+		if err != nil {
+			return nil, err
+		}
+		gitMetadataUploadOptions = append(
+			gitMetadataUploadOptions,
+			bufmodule.UploadWithCreateIfNotExist(createModuleVisibility, originRemote.HEADBranch()),
+		)
+	}
+	return gitMetadataUploadOptions, nil
+}
+
+func getGitMetadataLabelsUploadOptions(
+	ctx context.Context,
+	runner command.Runner,
+	env map[string]string,
+	input string,
+	gitCommitSha string,
+) (bufmodule.UploadOption, error) {
+	refs, err := git.GetRefsForGitCommitAndRemote(ctx, runner, env, input, gitOriginRemote, gitCommitSha)
+	if err != nil {
+		return nil, err
+	}
+	if len(refs) == 0 {
+		return nil, fmt.Errorf("no tags or branches found for HEAD, %s", gitCommitSha)
+	}
+	return bufmodule.UploadWithLabels(refs...), nil
+}
+
+// getGitMetadataSourceControlURLUploadOption takes a remote and git commit sha and makes
+// the best effort to construct a user-facing URL based on the hostname.
+//
+// If the remote hostname contains bitbucket (e.g. bitbucket.mycompany.com or bitbucket.org),
+// then it uses the route /commits for the git commit sha.
+//
+// If the remote hostname contains github (e.g. github.mycompany.com or github.com) or gitlab
+// (e.g. gitlab.mycompany.com or gitlab.com) then it uses the route /commit for the git
+// commit sha.
+func getGitMetadataSourceControlURLUploadOption(
```

Comment:
I think it would be good to move the building of source control URLs to the git package in order to keep the logic in here purely around upload. Perhaps a `SourceControlURL(commitSHA string)` on the `git.Remote` interface?

Reply:
Hmm, yeah, I went back and forth a little bit on where specifically this logic should live, since it's kind of "business-y" (rather than "generically git"), but also, the `RemoteKind` type already breaks that a little bit. So given that, it might make sense to put there.

It is a little strange to have this function take `commitSHA`, but I think that is reasonable. I'll make it clear that it just accepts the `commitSHA` as a string but doesn't do any validation against that string.

Reply:
Actually, in this case, we wouldn't need to expose `RemoteKind` at all. I think I'm going to unexport the enum and keep it since it is useful for keeping the parsing logic clean (and independent from the initial parsing of the URL).